### PR TITLE
Work in Progress for Admin center and Config editor

### PR DESCRIPTION
**Pull request by whoami-nr**

### CODE REVIEW

**Comment by rnikhil275** [comment]
I was working on a separate branch in my fork. From now on I will commit to master so that you can review all the commits in one plane i.e., here :)

**Comment by felipedaragon** [comment]
would not this make the admin center enabled by default and with a default password? if yes, for security reasons, I really would advise against having both, it enabled by default and a default password set https://github.com/sailorproject/sailor/pull/121/files#diff-aa9dd878052e45941a73be41206ba86aR17

**Comment by rnikhil275** [comment]
This is not the final config file. It won't be enabled by default for sure. The password also has to be set manually for it's usage.

**Comment by felipedaragon** [comment]
@rnikhil275 nice! 🙃

**Comment by Etiene** [comment]
@felipedaragon Yes, I told him to open the PR already so it would be easier for me to review the code in progress :)

**Comment by Etiene** [comment]
Don't forget to update automatic tests on the dev app once you are done, so travis CI tests start passing! If you need help with that, let me know :)

tests are located at the `test/dev-app`, you can cd there and run `sailor test`

File: src/sailor.lua [hunk 1]
```diff
@@ -107,13 +107,51 @@ end
...
+local function configedit(page)
+    local admin = require "sailor.admin"
+    local access = require "sailor.access"
+    if not access.is_guest() then
+        local src = admin.configedit()
+        src = lp.translate(src)
+        page:render('sailor/admin', {page=page},src)
+    else
+        return page:redirect('/admin')
+    end
```
Comment (Etiene) [inline src/sailor.lua]: These functions would be more organized if they moved to a different module

File: src/sailor/admin.lua [hunk 2]
```diff
@@ -0,0 +1,333 @@
...
+		<p>Configuration Edited Successfully</p>
+		<?lua
+	else
+		?>
+		<p>Error in Changing conf file</p>
+		<?lua
+	end
+
+
+
```
Comment (Etiene) [inline src/sailor/admin.lua]: This logic should not be in the same place as the html part

File: src/sailor/admin.lua [hunk 3]
```diff
@@ -0,0 +1,333 @@
...
+end
+?>
+
+
+
+	<h1> Config editor</h1><br>
```
Comment (Etiene) [inline src/sailor/admin.lua]: Indentation

File: src/sailor/admin.lua [hunk 4]
```diff
@@ -0,0 +1,333 @@
...
+
+
+<form method="post">
+<h3><u>Sailor Settings</u></h3>
+
+<table style="width:100%">
+
+<tr>
+<td>Admin Password</td>
+<td><input type="text"  name="admin_password" value="<%= conf.sailor.admin_password %>" /></td>
```
Comment (Etiene) [inline src/sailor/admin.lua]: Indentation

File: src/sailor/admin.lua [hunk 5]
```diff
@@ -0,0 +1,333 @@
...
+<td><input type="text"  name="lua_at_client" value="<%= conf.lua_at_client.vm %>" /></td>
+</tr>
+
+<tr>
+<td><u>Toggle Debug Mode</u></td>
+<td><input type="radio" name="debug" value="true">True
+<input type="radio" name="debug" value="false" checked>False</td>
+</tr>
+
+</table>
```
Comment (Etiene) [inline src/sailor/admin.lua]: Wouldn't it be better to put details in a table then loop through it to make the tables and the form fields? It would also be more maintainable

**Comment by Etiene** [comment]
I get stuck in a loop after I login in the admin center and try to access the autogen or the config edit when using Xavante

File: src/sailor/admin.lua (hunk 2 above)
Comment (rnikhil275) [inline src/sailor/admin.lua]: Yes. I agree with you. Should I make an admin folder inside /src/sailor and then put views and logic separately ? I think we have to modify sailor.lua to work with the same.

**Comment by rnikhil275** [comment]
It keeps getting redirected to the admin login page on Xavante probably because access.is_guest( ) always returns true in Xavante. This is the problem I was discussing with you.

**Comment by rnikhil275** [comment]
Regarding Indentations and moving functions outside sailor.lua, I will take care of them when I re-organize.

**Comment by rnikhil275** [comment]
Travis-CI fails now because the whole Autogen route/config thing has changed. Will update the tests. If I need help, will let you know :)

File: src/sailor/blank-app/conf/conf.lua [hunk 6]
```diff
@@ -11,19 +11,18 @@ local conf = {
 		default_error404 = 'error/404',
 		enable_autogen = false, -- default is false, should be true only in development environment
 		friendly_urls = false,
-		max_upload = 1024 * 1024,
```
Comment (Etiene) [inline src/sailor/blank-app/conf/conf.lua]: Why did you remove this line?

File: src/sailor/blank-app/conf/conf.lua [hunk 7]
```diff
@@ -11,19 +11,18 @@ local conf = {
...
+		enable_admin=false, -- set true only in development. Set password below too.
+		admin_password=""
 	},
 
 	db = {
 		development = { -- current environment
-			driver = 'mysql',
-			host = '',
-			user = '',
-			pass = '',
```
Comment (Etiene) [inline src/sailor/blank-app/conf/conf.lua]: Why did you modify this?

Comment (rnikhil275) [inline src/sailor/blank-app/conf/conf.lua]: Yeah. I expected that and I have been meaning to include it in the commit message as a breaking change. The name attribute of the fields are generated with a loop(in dashboard.lp). The Mail settings and the db settings had common fields such as user & pass. This created a nested table in page.POST when I tried looping through it in the controller. I wanted to generate the config file with a loop but I had to add strings in between(string1,string2 etc). So I made the file writing part manual with each value written separately. One thing that can be done is that, the db fields can be reverted back and I will change the smtp fields.

File: src/sailor/blank-app/conf/conf.lua (hunk 6 above)
Comment (rnikhil275) [inline src/sailor/blank-app/conf/conf.lua]: I fixed it. Removed it accidentally.

File: src/sailor/blank-app/conf/conf.lua (hunk 7 above)
Comment (rnikhil275) [inline src/sailor/blank-app/conf/conf.lua]: I made a commit reflecting the change too

File: src/sailor/admin.lua (hunk 5 above)
Comment (rnikhil275) [inline src/sailor/admin.lua]: I made this change. But I couldn't loop the similar way while writing into the config file. I mentioned it in detail here: https://github.com/sailorproject/sailor/pull/121#discussion_r68693139

**Comment by rnikhil275** [comment]
With the last commit, the dashboard is now functional with autogen functions and the config editor. Please review it and let me know if I need to fix anything. I will start on the tests next

**Comment by rnikhil275** [comment]
Category.lua tests are failing which I don't understand why. I will remove the old autogen.lua test file now that it's inside the admin.lua test file.

**Comment by rnikhil275** [comment]
The coverage possibly decreased because the tests are incomplete.

**Comment by Etiene** [comment]
closing due to lack of time to readdress this
